File: packages/MeeshySDK/Sources/MeeshySDK/Utils/MimeTypeResolver.swift
import Foundation

/// Single source of truth pour les conversions extension de fichier ↔ mimeType.
///
/// Avant ce type, trois sites au moins dupliquaient cette table :
/// `ConversationView+AttachmentHandlers.swift` (la version complète),
/// `FeedView+Attachments.swift:468` (un sous-ensemble réduit avec un bug `docx`
/// mappé sur `application/msword`), et `FeedView+Attachments.swift:1225` (un
/// sous-ensemble encore plus restreint). Le résolveur unifie ces tables et
/// fournit aussi la direction inverse (`mimeType → preferred extension`) pour
/// les services qui doivent matérialiser un fichier à partir d'un mime
/// (notification rich push, partage UTI, etc.).
///
/// La table est volontairement statique : les mimes "à la mode" évoluent peu,
/// pas de raison de pousser ça en config dynamique.
public enum MimeTypeResolver {

    /// Convertit une extension de fichier (avec ou sans `.` initial,
    /// case-insensitive) vers son mime type canonique. Retourne
    /// `application/octet-stream` pour les extensions inconnues — ce qui est
    /// la valeur RFC-compliant pour "bytes opaques".
    public static func mimeType(forExtension ext: String) -> String {
        let normalized = normalizeExtension(ext)
        return forwardTable[normalized] ?? "application/octet-stream"
    }

    /// Convertit l'extension du `pathExtension` d'une URL vers son mime type.
    /// Raccourci pour les sites qui ont une URL sous la main (file picker,
    /// upload depuis le caméra, etc.).
    public static func mimeType(forURL url: URL) -> String {
        mimeType(forExtension: url.pathExtension)
    }

    /// Convertit l'extension d'un nom de fichier vers son mime type.
    /// `report.xlsx` → `application/vnd.openxmlformats-officedocument.spreadsheetml.sheet`.
    /// Un nom sans extension renvoie le fallback `application/octet-stream`.
    public static func mimeType(forFilename filename: String) -> String {
        guard let lastDot = filename.lastIndex(of: "."),
              lastDot != filename.index(before: filename.endIndex) else {
            return "application/octet-stream"
        }
        let ext = filename[filename.index(after: lastDot)...]
        return mimeType(forExtension: String(ext))
    }

    /// Direction inverse : depuis un mime type, propose l'extension la plus
    /// naturelle pour matérialiser un fichier (sans le `.`). Retourne `nil`
    /// pour les mimes non couverts par la table (le caller décide alors d'un
    /// fallback générique, p.ex. `m4a` pour un audio inconnu).
    ///
    /// Cas-insensitive sur l'input.
    public static func preferredExtension(for mimeType: String) -> String? {
        let normalized = mimeType.lowercased()
        return reverseTable[normalized]
    }

    // MARK: - Tables

    private static func normalizeExtension(_ ext: String) -> String {
        var trimmed = ext.lowercased()
        if trimmed.hasPrefix(".") { trimmed.removeFirst() }
        return trimmed
    }

    /// Table extension → mime. Toute extension absente tombe sur le fallback.
    /// Les couples synonymes (jpg/jpeg, mp4/m4v, heic/heif, ogg/oga, etc.)
    /// pointent vers le même mime canonique.
    private static let forwardTable: [String: String] = [
        // Images
        "jpg":  "image/jpeg",
        "jpeg": "image/jpeg",
        "png":  "image/png",
        "gif":  "image/gif",
        "webp": "image/webp",
        "heic": "image/heic",
        "heif": "image/heic",
        "svg":  "image/svg+xml",
        "bmp":  "image/bmp",
        "tiff": "image/tiff",
        "tif":  "image/tiff",
        // Video
        "mp4":  "video/mp4",
        "m4v":  "video/mp4",
        "mov":  "video/quicktime",
        "avi":  "video/x-msvideo",
        "mkv":  "video/x-matroska",
        "webm": "video/webm",
        // Audio
        "mp3":  "audio/mpeg",
        "m4a":  "audio/mp4",
        "aac":  "audio/mp4",
        "wav":  "audio/wav",
        "ogg":  "audio/ogg",
        "oga":  "audio/ogg",
        "flac": "audio/flac",
        "wma":  "audio/x-ms-wma",
        // Documents
        "pdf":  "application/pdf",
        "doc":  "application/msword",
        "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "xls":  "application/vnd.ms-excel",
        "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "ppt":  "application/vnd.ms-powerpoint",
        "pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "rtf":  "application/rtf",
        // iWork
        "pages":   "application/x-iwork-pages-sffpages",
        "numbers": "application/x-iwork-numbers-sffnumbers",
        "keynote": "application/x-iwork-keynote-sffkey",
        // Text & Code
        "txt":      "text/plain",
        "log":      "text/plain",
        "csv":      "text/csv",
        "html":     "text/html",
        "htm":      "text/html",
        "css":      "text/css",
        "md":       "text/markdown",
        "markdown": "text/markdown",
        "json":     "application/json",
        "xml":      "application/xml",
        "js":       "application/javascript",
        "ts":       "application/typescript",
        "py":       "text/x-python",
        "swift":    "text/x-swift",
        // Archives
        "zip":  "application/zip",
        "rar":  "application/x-rar-compressed",
        "7z":   "application/x-7z-compressed",
        "tar":  "application/x-tar",
        "gz":   "application/gzip",
        "gzip": "application/gzip",
    ]

    /// Table inverse mime → extension préférée. Pour les mimes ayant plusieurs
    /// extensions (jpeg : jpg/jpeg, mp4 : mp4/m4v, etc.) on retient la plus
    /// courante (jpg, mp4, m4a, ogg, tiff, md, gz).
    private static let reverseTable: [String: String] = [
        "image/jpeg": "jpg",
        "image/png":  "png",
        "image/gif":  "gif",
        "image/webp": "webp",
        "image/heic": "heic",
        "image/svg+xml": "svg",
        "image/bmp":  "bmp",
        "image/tiff": "tiff",
        "video/mp4":         "mp4",
        "video/quicktime":   "mov",
        "video/x-msvideo":   "avi",
        "video/x-matroska":  "mkv",
        "video/webm":        "webm",
        "audio/mpeg":     "mp3",
        "audio/mp4":      "m4a",
        "audio/wav":      "wav",
        "audio/ogg":      "ogg",
        "audio/flac":     "flac",
        "audio/x-ms-wma": "wma",
        "application/pdf":     "pdf",
        "application/msword":  "doc",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document": "docx",
        "application/vnd.ms-excel":  "xls",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": "xlsx",
        "application/vnd.ms-powerpoint": "ppt",
        "application/vnd.openxmlformats-officedocument.presentationml.presentation": "pptx",
        "application/rtf": "rtf",
        "application/x-iwork-pages-sffpages":     "pages",
        "application/x-iwork-numbers-sffnumbers": "numbers",
        "application/x-iwork-keynote-sffkey":     "keynote",
        "text/plain":      "txt",
        "text/csv":        "csv",
        "text/html":       "html",
        "text/css":        "css",
        "text/markdown":   "md",
        "text/x-python":   "py",
        "text/x-swift":    "swift",
        "application/json":       "json",
        "application/xml":        "xml",
        "application/javascript": "js",
        "application/typescript": "ts",
        "application/zip":              "zip",
        "application/x-rar-compressed": "rar",
        "application/x-7z-compressed":  "7z",
        "application/x-tar":            "tar",
        "application/gzip":             "gz",
    ]
}

File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/MimeTypeResolver.kt
package me.meeshy.sdk.model

/**
 * Single source of truth for file-extension ↔ mimeType conversions. Faithful port
 * of the iOS `MimeTypeResolver` (MeeshySDK `Utils/MimeTypeResolver.swift`) — the
 * same forward (extension → mime) and reverse (mime → preferred extension) tables,
 * including the synonym couples (jpg/jpeg, mp4/m4v, heic/heif, ogg/oga …) that
 * point at one canonical mime.
 *
 * A pure, stateless building block: it takes opaque strings and returns strings,
 * with no product decision baked in — so it belongs in `:core:model`. The
 * "which picker / when to upload" orchestration lives in the feature layer.
 */
object MimeTypeResolver {

    private const val OCTET_STREAM = "application/octet-stream"

    /**
     * Converts a file extension (with or without a leading `.`, case-insensitive)
     * to its canonical mime type. Returns [OCTET_STREAM] — the RFC-compliant value
     * for "opaque bytes" — for unknown extensions.
     */
    fun mimeTypeForExtension(ext: String): String =
        FORWARD_TABLE[normalizeExtension(ext)] ?: OCTET_STREAM

    /**
     * Converts the last extension of a filename to its mime type. A name with no
     * extension — or one ending in a bare `.` — falls back to [OCTET_STREAM].
     */
    fun mimeTypeForFilename(fileName: String): String {
        val lastDot = fileName.lastIndexOf('.')
        if (lastDot < 0 || lastDot == fileName.length - 1) return OCTET_STREAM
        return mimeTypeForExtension(fileName.substring(lastDot + 1))
    }

    /**
     * Reverse direction: from a mime type, the most natural extension (without the
     * `.`) to materialise a file. Returns `null` for mimes absent from the table so
     * the caller can pick its own generic fallback. Case-insensitive on the input.
     */
    fun preferredExtensionForMime(mime: String): String? =
        REVERSE_TABLE[mime.lowercase()]

    /**
     * Chooses the best mime for a picked attachment: a meaningful [declaredType]
     * (what the platform's content resolver reported — non-blank and not the opaque
     * [OCTET_STREAM]) wins; otherwise the [fileName]'s extension is consulted. This
     * mirrors the iOS picker, which always derives the mime from the URL extension,
     * but prefers a real declared content-type when Android supplies one.
     */
    fun resolve(declaredType: String?, fileName: String): String {
        val declared = declaredType?.trim().orEmpty()
        if (declared.isNotEmpty() && !declared.equals(OCTET_STREAM, ignoreCase = true)) {
            return declared
        }
        return mimeTypeForFilename(fileName)
    }

    private fun normalizeExtension(ext: String): String =
        ext.lowercase().removePrefix(".")

    private val FORWARD_TABLE: Map<String, String> = mapOf(
        // Images
        "jpg" to "image/jpeg",
        "jpeg" to "image/jpeg",
        "png" to "image/png",
        "gif" to "image/gif",
        "webp" to "image/webp",
        "heic" to "image/heic",
        "heif" to "image/heic",
        "svg" to "image/svg+xml",
        "bmp" to "image/bmp",
        "tiff" to "image/tiff",
        "tif" to "image/tiff",
        // Video
        "mp4" to "video/mp4",
        "m4v" to "video/mp4",
        "mov" to "video/quicktime",
        "avi" to "video/x-msvideo",
        "mkv" to "video/x-matroska",
        "webm" to "video/webm",
        // Audio
        "mp3" to "audio/mpeg",
        "m4a" to "audio/mp4",
        "aac" to "audio/mp4",
        "wav" to "audio/wav",
        "ogg" to "audio/ogg",
        "oga" to "audio/ogg",
        "flac" to "audio/flac",
        "wma" to "audio/x-ms-wma",
        // Documents
        "pdf" to "application/pdf",
        "doc" to "application/msword",
        "docx" to "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "xls" to "application/vnd.ms-excel",
        "xlsx" to "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "ppt" to "application/vnd.ms-powerpoint",
        "pptx" to "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "rtf" to "application/rtf",
        // iWork
        "pages" to "application/x-iwork-pages-sffpages",
        "numbers" to "application/x-iwork-numbers-sffnumbers",
        "keynote" to "application/x-iwork-keynote-sffkey",
        // Text & Code
        "txt" to "text/plain",
        "log" to "text/plain",
        "csv" to "text/csv",
        "html" to "text/html",
        "htm" to "text/html",
        "css" to "text/css",
        "md" to "text/markdown",
        "markdown" to "text/markdown",
        "json" to "application/json",
        "xml" to "application/xml",
        "js" to "application/javascript",
        "ts" to "application/typescript",
        "py" to "text/x-python",
        "swift" to "text/x-swift",
        // Archives
        "zip" to "application/zip",
        "rar" to "application/x-rar-compressed",
        "7z" to "application/x-7z-compressed",
        "tar" to "application/x-tar",
        "gz" to "application/gzip",
        "gzip" to "application/gzip",
    )

    private val REVERSE_TABLE: Map<String, String> = mapOf(
        "image/jpeg" to "jpg",
        "image/png" to "png",
        "image/gif" to "gif",
        "image/webp" to "webp",
        "image/heic" to "heic",
        "image/svg+xml" to "svg",
        "image/bmp" to "bmp",
        "image/tiff" to "tiff",
        "video/mp4" to "mp4",
        "video/quicktime" to "mov",
        "video/x-msvideo" to "avi",
        "video/x-matroska" to "mkv",
        "video/webm" to "webm",
        "audio/mpeg" to "mp3",
        "audio/mp4" to "m4a",
        "audio/wav" to "wav",
        "audio/ogg" to "ogg",
        "audio/flac" to "flac",
        "audio/x-ms-wma" to "wma",
        "application/pdf" to "pdf",
        "application/msword" to "doc",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document" to "docx",
        "application/vnd.ms-excel" to "xls",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" to "xlsx",
        "application/vnd.ms-powerpoint" to "ppt",
        "application/vnd.openxmlformats-officedocument.presentationml.presentation" to "pptx",
        "application/rtf" to "rtf",
        "application/x-iwork-pages-sffpages" to "pages",
        "application/x-iwork-numbers-sffnumbers" to "numbers",
        "application/x-iwork-keynote-sffkey" to "keynote",
        "text/plain" to "txt",
        "text/csv" to "csv",
        "text/html" to "html",
        "text/css" to "css",
        "text/markdown" to "md",
        "text/x-python" to "py",
        "text/x-swift" to "swift",
        "application/json" to "json",
        "application/xml" to "xml",
        "application/javascript" to "js",
        "application/typescript" to "ts",
        "application/zip" to "zip",
        "application/x-rar-compressed" to "rar",
        "application/x-7z-compressed" to "7z",
        "application/x-tar" to "tar",
        "application/gzip" to "gz",
    )
}
